include/lock/optimistic_lock.hpp

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2021 Database Group, Nagoya University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CPP_UTILITY_OPTIMISTIC_LOCK_HPP
+#define CPP_UTILITY_OPTIMISTIC_LOCK_HPP
+
+#include <atomic>
+#include <chrono>
+#include <thread>
+
+#include "common.hpp"
+
+namespace dbgroup::lock
+{
+class OptimisticLock
+{
+ public:
+  /*####################################################################################
+   * Public constructors and assignment operators
+   *##################################################################################*/
+
+  constexpr OptimisticLock() = default;
+
+  OptimisticLock(const OptimisticLock&) = delete;
+  OptimisticLock(OptimisticLock&&) = delete;
+
+  auto operator=(const OptimisticLock&) -> OptimisticLock& = delete;
+  auto operator=(OptimisticLock&&) -> OptimisticLock& = delete;
+
+  /*####################################################################################
+   * Public destructors
+   *##################################################################################*/
+
+  ~OptimisticLock() = default;
+
+  /*####################################################################################
+   * Public getters
+   *##################################################################################*/
+  /**
+   * @brief Get the Version object
+   *
+   * @return std::pair<uint64_t, bool>
+   */
+  [[nodiscard]] auto
+  GetVersion() const  //
+      -> uint64_t
+  {
+    while (true) {
+      auto expected = lock_.load(std::memory_order_acquire);
+      for (size_t i = 0; i < kRetryNum; ++i) {
+        if ((expected & kXLock) == 0) return expected & kAllBitsMask;
+
+        CPP_UTILITY_SPINLOCK_HINT
+        expected = lock_.load(std::memory_order_acquire);
+      }
+
+      std::this_thread::sleep_for(kShortSleep);
+    }
+  }
+
+  /**
+   * @brief Check whether the version after reading is equal to the version before reading
+   *
+   * @return true
+   * @return false
```

Comment:
```suggestion
   * @param expected an expected version value.
   * @retval true if the given version value is the same as a current one.
   * @retval false otherwise.
```